### PR TITLE
Add a lightweight e2e test setup and tests

### PR DESCRIPTION
**Pull request by kienstra**
#### Changes
Adds an e2e test that runs in a headless browser and:
1. Activates Block Lab
2. Creates a new Block Lab block with a Text field
3. Create a new post
4. Adds that blocks to the post
5. Enters text in the text field
6. Clicks away from the block so the `<ServerSideRender>` displays
7. Verifies that the blocks looks good, using the PHP template that this PR adds. This tests that `block_value()` and `block_field()` have the correct output.

This uses the e2e test scripts from Gutenberg, minimizing the amount of code in this PR.

The test is simple because e2e tests can be really flaky. 

This should make it easier to merge `dependabot` PRs, knowing that the plugin works at a high level. 

For fine-grained JS tests, the [component tests](https://github.com/getblocklab/block-lab/tree/ad4981a3351c6136cc4d55b4cb2e21a182a54a96/js/blocks/components/test) are better, and have been reliable so far.

#### Testing instructions
```bash
npm install # might be necessary to remove node_modules first
npm run build
npm run env start
npm run test:e2e # To see the test running, do npm run test:e2e:interactive
npm run env stop # Do this whenever, but you probably don't want the container running all of the time.
```
Because this updates most of the packages in `package.json`, this could use full regression testing.

### CODE REVIEW

**Comment by kienstra** [comment]
Hi @lukecarbis,
Could you please review this when you have a chance?

This test runs in about 10 seconds, on the `npm run test:js` command. It could be a good alternative to e2e testing in a headless browser, which can be flaky and take minutes.

Thanks, Luke!

**Comment by lukecarbis** [comment]
Takes a little longer than 10 seconds for me, and the tests don't pass. Something wrong on my end?

My process:
1. `git checkout add/e2e-testing-editor`
2. `git pull`
3. `npm install`
4. `npm run build`
5. `npm run test:js`

Screen Shot 2020-04-17 at 5 54 02 am

**Comment by kienstra** [comment]
Hm...

**Comment by kienstra** [comment]
Hi @lukecarbis,
Thanks for testing this. Those commands look right, and the tests passed in my local with them.

This shouldn't be necessary, but could you try:
```sh
rm -rf node_modules
npm install
npm run test:js
```

**Comment by kienstra** [comment]
It's strange, that failed test [above](https://github.com/getblocklab/block-lab/pull/525#issuecomment-614862633) is from an existing component test.

**Comment by lukecarbis** [comment]
Did the following:

```
rm -rf node_modules
npm install
npm run build
npm run test:js
```

Still got errors, though a bit different this time: https://pastebin.com/GTAque2P

**Comment by kienstra** [comment]
OK, thanks for posting those. 

What version of Node is running?

**Comment by kienstra** [comment]
Also, do the tests pass on the `develop` branch?

**Comment by lukecarbis** [comment]
Ah – good thinking. No, they don't pass in `develop`. I get this:

```
> wp-scripts test-unit-js --config=tests/js/jest.config.js

..
  ● FetchInput › should only have the error class if there are no results after focusing

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Warning: You seem to have overlapping act() calls, this is not supported. Be sure to await previous act() calls before making a new one. "]

      at Object.assertExpectedCalls (node_modules/@wordpress/jest-console/build/@wordpress/jest-console/src/index.js:36:4)

....  console.warn node_modules/jest-mock/build/index.js:866
    wp.nux is deprecated. Note: wp.components.Guide can be used to show a user guide.

.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! block-lab@1.5.4 test:js: `wp-scripts test-unit-js --config=tests/js/jest.config.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the block-lab@1.5.4 test:js script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lukecarbis/.npm/_logs/2020-04-18T21_43_19_876Z-debug.log
```

**Comment by kienstra** [comment]
Hi @lukecarbis,
Thanks for testing that. Does the latest commit help?

If not, which version of Node is running?

**Comment by lukecarbis** [comment]
That commit seems to fix the `FetchInput`, error, but I still have the `TextBlock` error. (See the pastebin above).

```
$ npm -v
> 6.14.4
$ node --version
> v13.3.0
```

**Comment by kienstra** [comment]
Thanks, I can reproduce the failed `test:js` now.

**Comment by lukecarbis** [comment]
Let me know when you're ready for me to take another look. :)

**Comment by kienstra** [comment]
OK, thanks! I think it's in pretty good shape, but I might add another test if time permits.

**Comment by kienstra** [comment]
These tests look to be reliable so far. The [test-js](https://travis-ci.com/github/getblocklab/block-lab/jobs/324156870) [builds](https://travis-ci.com/github/getblocklab/block-lab/jobs/324156889) passed 10 times in a row.

There could be flaky failures when updating packages, like if `@wordpress/editor` has changes of class names that this depends on. But development PRs shouldn't be flaky.

**Comment by lukecarbis** [comment]
Still getting an error. This one is new, though:

```
> block-lab@1.5.5 test:js /Users/lukecarbis/Sites/block-lab/app/public/wp-content/plugins/block-lab
> wp-scripts test-unit-js --config=tests/js/jest.config.js

.  console.warn node_modules/jest-environment-jsdom-sixteen/node_modules/jest-mock/build/index.js:814
    wp.nux is deprecated. Note: wp.components.Guide can be used to show a user guide.

..  console.warn node_modules/jest-environment-jsdom-sixteen/node_modules/jest-mock/build/index.js:814
    wp.nux is deprecated. Note: wp.components.Guide can be used to show a user guide.

.  console.warn node_modules/jest-environment-jsdom-sixteen/node_modules/jest-mock/build/index.js:814
    wp.nux is deprecated. Note: wp.components.Guide can be used to show a user guide.

.....
  ● RepeaterBlock › displays the repeater block in the inserter and the block has the expected values

    TypeError: Cannot read property 'calls' of undefined

      at Object.<anonymous> (node_modules/@wordpress/jest-console/build/@wordpress/jest-console/src/matchers.js:16:25)
      at Object.throwingMatcher [as toHaveErrored] (node_modules/expect/build/index.js:342:33)

**Comment by kienstra** [comment]
OK, thanks. I'll look at that, it's strange that one of our environments doesn't have an error, and one does.

**Comment by kienstra** [comment]
The tests fail even after a fresh` npm install`, right?

**Comment by kienstra** [comment]
All good, I have some debugging to do with the latest commits.

**Comment by kienstra** [comment]
There might be a need for a `<BlocksProvider>` to make the `'core/blocks'` store available, like how the [<BlockEditorProvider>](https://github.com/getblocklab/block-lab/pull/525/files#diff-354aac6e1a663d62c4625d3b22a0e6b6R58) exposes the `'core/block-editor'`. 

It looks like the [errors](https://travis-ci.com/github/getblocklab/block-lab/jobs/324768999#L702) are from `select( 'core/blocks' )` being `null`:

```
TypeError: Cannot read property 'getBlockTypes' of null
```

File: js/blocks/components/test/fetch-input.js [hunk 1]
```diff
@@ -42,39 +39,56 @@ const baseProps = {
...
+			apiFetch.mockImplementationOnce(
+				() => new Promise( ( resolve ) => resolve( apiResults ) )
+			);
 			const { input } = setup( baseProps );
 			fireEvent.focus( input );
 
-			await waitForDomChange( { container: input } ).then( () => {
-				expect( input.classList.contains( 'text-control__error' ) ).toStrictEqual( expected );
-			} );
+			waitFor( () =>
```
Comment (kienstra) [inline js/blocks/components/test/fetch-input.js:73]: This isn't related to e2e testing, but it fixes a failed test when I updated the `package.json` script versions.

File: package.json [hunk 2]
```diff
@@ -18,72 +18,85 @@
...
-    "@wordpress/url": "2.13.0",
-    "autoprefixer": "9.7.3",
+    "@wordpress/html-entities": "2.7.0",
+    "@wordpress/i18n": "3.12.0",
+    "@wordpress/jest-console": "3.6.0",
+    "@wordpress/keycodes": "2.12.0",
+    "@wordpress/scripts": "9.0.0",
+    "@wordpress/server-side-render": "1.11.0",
+    "@wordpress/url": "2.14.0",
+    "autoprefixer": "9.7.6",
```
Comment (kienstra) [inline package.json:46]: I went a little overboard in updating all of these `@wordpress` packages, but figured we should test the latest versions.

File: package.json [hunk 3]
```diff
@@ -18,72 +18,85 @@
...
-    "@wordpress/server-side-render": "1.5.0",
-    "@wordpress/url": "2.13.0",
-    "autoprefixer": "9.7.3",
+    "@wordpress/html-entities": "2.7.0",
+    "@wordpress/i18n": "3.12.0",
+    "@wordpress/jest-console": "3.6.0",
+    "@wordpress/keycodes": "2.12.0",
+    "@wordpress/scripts": "9.0.0",
+    "@wordpress/server-side-render": "1.11.0",
+    "@wordpress/url": "2.14.0",
```
Comment (kienstra) [inline package.json:45]: I went a little overboard in updating all of these `@wordpress` packages, but figured we should test the latest versions.

**Comment by kienstra** [comment]
Do you use the [bundled Docker setup](https://github.com/getblocklab/block-lab/tree/ad4981a3351c6136cc4d55b4cb2e21a182a54a96/tests/docker)? This PR uses the Docker setup from Gutenberg, so maybe we should delete the existing setup if we're not using it. 

It doesn't hurt to have the other one, it just might be confusing to have 2 Docker setups.

File: .travis.yml [hunk 4]
```diff
@@ -32,12 +32,31 @@ matrix:
       env: WP_TRAVISCI=lint
     - php: 7.3
       env: WP_TRAVISCI=test-js
+    - php: 7.3
+      env: WP_TRAVISCI=test-e2e
 
 install:
+  - set -e
```
Comment (kienstra) [inline .travis.yml:39]: This [means that](https://ss64.com/bash/set.html) if a command exits with non-zero, the build will fail. For example, if `npm run env start` fails, the e2e test will fail also. 

So it makes sense to let it display the error, and the build.

File: .travis.yml [hunk 5]
```diff
@@ -56,12 +72,15 @@ before_script:
...
     if [[ "$WP_TRAVISCI" == "lint" ]] ; then
+      set -e
       composer global require wp-coding-standards/wpcs
       phpcs --config-set installed_paths $HOME/.composer/vendor/wp-coding-standards/wpcs
       npm run lint:js
+      set +e
     fi
 
 script:
+  - set -e
```
Comment (kienstra) [inline .travis.yml]: This [means that](https://ss64.com/bash/set.html) if a command exits with non-zero, the build will fail. 

Before, there was a problem with an e2e failing, but the build still passing.

File: .eslintrc [hunk 6]
```diff
@@ -32,5 +32,30 @@
       }
     ],
     "jsx-a11y/alt-text": "warn"
-  }
+  },
+  "overrides": [
```
Comment (kienstra) [inline .eslintrc:36]: Linting rules for the `tests/e2e/` files. Without these, there are ESLint errors from using the `page` global, etc...

File: .npmrc [hunk 7]
```diff
@@ -0,0 +1 @@
+save-exact = true
```
Comment (kienstra) [inline .npmrc:1]: Ensures that on adding new `package.json` files, it saves the exact version, not something with a caret like `^3.5.3`. 

This should help ensure our builds are the same, even in our different locals.

File: package.json [hunk 8]
```diff
@@ -18,72 +18,86 @@
...
-    "sass-loader": "^8.0.2",
-    "style-loader": "^1.0.2",
-    "uglifyjs-webpack-plugin": "^2.2.0",
-    "webpack": "^4.41.3",
-    "webpack-cli": "3.3.10"
+    "react": "16.13.1",
+    "react-dom": "16.13.1",
+    "sass-loader": "8.0.2",
+    "style-loader": "1.0.2",
+    "uglifyjs-webpack-plugin": "2.2.0",
```
Comment (kienstra) [inline package.json:75]: The leading `^` is removed, as it should help our builds be similar, even in our different locals.

It forces the exact version, and doesn't allow upgrading to minor versions.

File: .nvmrc [hunk 9]
```diff
@@ -0,0 +1 @@
+lts/*
```
Comment (kienstra) [inline .nvmrc:1]: This sets the Node version to LTS.

File: js/admin.block-post.js [hunk 10]
```diff
@@ -1,9 +1,8 @@
 /**
  * Used for editing Blocks.
  *
- * @package   Block_Lab
  * @copyright Copyright(c) 2020, Block Lab
- * @license http://opensource.org/licenses/GPL-2.0 GNU General Public License, version 2 (GPL-2.0)
+ * @license GPL-2.0-only
```
Comment (kienstra) [inline js/admin.block-post.js:5]: Changed in response to a linting warning

File: package.json [hunk 11]
```diff
@@ -18,72 +18,84 @@
   "bugs": {
     "url": "https://github.com/getblocklab/block-lab/issues"
   },
-  "dependencies": {
-    "classnames": "^2.2.6"
```
Comment (kienstra) [inline package.json:22]: Moved to `devDependencies`

File: .travis.yml [hunk 12]
```diff
@@ -12,6 +9,9 @@ cache:
   directories:
     - $HOME/.composer/cache
 
+services:
+  - mysql
+
```
Comment (kienstra) [inline .travis.yml:14]: A lot of changes to this file were needed for the e2e tests. The upside is that there isn't much forked code, like a `docker-compose.yml` file.

File: .wp-env.json [hunk 13]
```diff
@@ -0,0 +1,4 @@
+{
+	"core": null,
+	"plugins": [ ".", "./tests/e2e/plugins/testing-blocks" ]
```
Comment (kienstra) [inline .wp-env.json:3]: Points the local environment to the Block Lab plugin, and a testing plugin, so they can run in the environment.

File: package.json [hunk 14]
```diff
@@ -18,72 +18,84 @@
...
+    "@wordpress/compose": "3.14.0",
+    "@wordpress/core-data": "2.15.0",
+    "@wordpress/data": "4.17.0",
+    "@wordpress/dependency-extraction-webpack-plugin": "2.6.0",
+    "@wordpress/dom": "2.9.0",
+    "@wordpress/e2e-test-utils": "4.6.0",
+    "@wordpress/e2e-tests": "1.15.0",
+    "@wordpress/editor": "9.15.0",
+    "@wordpress/element": "2.13.1",
+    "@wordpress/env": "1.3.0",
```
Comment (kienstra) [inline package.json:37]: Most of the diff in `package.json` is updating existing packages, or removing the `^` from the version. There are some new packages, though.

**Comment by kienstra** [comment]
**Request for another review**

Hi @lukecarbis,
Good to see you earlier. Could you please review this again when you have a chance? It's completely different from before.

The [top comment](https://github.com/getblocklab/block-lab/pull/525#issue-399419472) is updated.

This test is pretty basic, so that ideally this won't be very flaky. 

Thanks, Luke!

**Comment by mindctrl** [comment]
@kienstra these tests run and pass for me. Takes a little under 10s to complete.

**Comment by kienstra** [comment]
@mindctrl, thanks so much for testing this. Good to hear they work well for you.

**Comment by lukecarbis** [comment]
@kienstra I must be doing something wrong. After run-in `npm install` and `npm run build`, if I run `npm run env start` I get this error:

```
> block-lab@1.5.5 env /Users/lukecarbis/Sites/block-lab/app/public/wp-content/plugins/block-lab
> wp-env "start"

✖ spawn docker-compose ENOENT
Error: spawn docker-compose ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:266:19)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn docker-compose',
  path: 'docker-compose',
  spawnargs: [
    '-f',
    '/Users/lukecarbis/.wp-env/63dc7968e06288cf8a67f40dd88463dd/docker-compose.yml',
    'down'
  ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! block-lab@1.5.5 env: `wp-env "start"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the block-lab@1.5.5 env script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by kienstra** [comment]
Hi @lukecarbis,
Thanks for testing this. I don't think you're making a mistake, maybe it's just a difference in the computers.

Are there any updates available for Docker on your computer?

**Comment by lukecarbis** [comment]
I don't have docker installed on my machine.

**Review by lukecarbis — APPROVED** [review]

**Comment by kienstra** [comment]
Ah, OK. I don't like running a virtual environment, I just installed Docker on this machine last week :)

It usually shouldn't be needed to run the e2e tests when developing, unless there's a failure in Travis that you need to reproduce.

**Pull request merged**